samples/siwx91x_ota/src/main.c

```diff
@@ -0,0 +1,875 @@
+
+	/* Wait for HTTP response to be fully received */
+	if (k_sem_take(&app_ctx_st->ota_sem, K_MSEC(FW_CHUNK_DOWNLOAD_TIMEOUT)) != 0) {
+		printf("Firmware download failed or Semaphore timed out\n");
+		return -1;
+	}
+
+	/* Check if there was an error in the HTTP response */
+	if (http_response_error) {
+		printf("Error occurred during HTTP download, skipping processing\n");
+		return -1;
+	}
+	req_start = start;
+
+	/* Process all queued data chunks */
+	while (k_msgq_get(&http_data_q, &msg_c, K_NO_WAIT) == 0) {
+		start = start + msg_c.length;
+		status = ota_load_firmware(msg_c.buffer, msg_c.length);
+		if (status != SL_STATUS_OK) {
+			printf("Fw load failed for requested range=%d-%d\n", req_start, end);
+			return -1;
+		}
+	}
+	printf("Fw load success for requested range=%d-%d\n", req_start, end);
+	return 0;
+}
+
+/*
+ * @brief Main OTA application state machine
+ */
+static void ota_application_start(struct app_ctx *app_ctx_st)
+{
+	int ret = 0;
+	int firmware_status = 0;
+	char *ssid = CONFIG_OTA_WIFI_SSID;
+	char *pwd = CONFIG_OTA_WIFI_PSK;
+	sl_wifi_firmware_version_t version = {0};
+	bool ota_complete = false;
+
+	printf("HTTP Client Application Started\n");
+
+	/* Main state machine loop */
+	while (1) {
+		switch (state) {
+		case OTA_STATE_INIT:
+			printf("Initializing OTA update process\n");
+			/* Get initial firmware version */
+			ret = sl_wifi_get_firmware_version(&version);
+			if (ret == SL_STATUS_OK) {
+				print_firmware_version(&version);
+			} else {
+				printf("Failed to get firmware version: 0x%x\n", ret);
+			}
+			state = OTA_STATE_SCAN;
+			break;
+
+		case OTA_STATE_SCAN:
+			ret = wifi_start_scan(app_ctx_st);
+			if (ret < 0) {
+				state = OTA_STATE_SCAN;
+			}
+			break;
+		case OTA_STATE_CONNECT:
+			ret = wifi_connect(ssid, pwd, CONFIG_OTA_WIFI_SECURITY_TYPE, app_ctx_st);
+			if (ret < 0) {
+				state = OTA_STATE_CONNECT;
+			}
+			break;
+
+		case OTA_STATE_IP_CONFIG:
+			ret = configure_ip(app_ctx_st);
+			state = (ret == 0) ? OTA_STATE_SERVER_CONNECT : OTA_STATE_IP_CONFIG;
+			break;
+
+		case OTA_STATE_SERVER_CONNECT:
+			if (!ipv6_addr_config_done) {
+				k_sleep(K_MSEC(10));
+			}
+			ret = connect_to_ota_server(&sock);
+			if (ret < 0) {
+				/* Retry on timeout */
+				k_sleep(K_MSEC(1000));
+				state = OTA_STATE_SERVER_CONNECT;
+				break;
+			}
+			state = OTA_STATE_DOWNLOAD;
+			break;
+
+		case OTA_STATE_DOWNLOAD:
+			ret = download_firmware_chunk(sock, start, end, app_ctx_st);
+			if (ret < 0) {
+				if (sock >= 0) {
+					zsock_close(sock);
+					sock = -1;
+				}
+				k_sleep(K_MSEC(1000));
+				state = OTA_STATE_SERVER_CONNECT;
+				break;
+			}
+			break;
+
+		case OTA_STATE_PROCESS:
+			firmware_status = process_firmware_chunk(app_ctx_st);
+			if (firmware_status < 0) {
+				printf("OTA update failed\n");
+				/* Clean up resources */
+				if (sock >= 0) {
+					zsock_close(sock);
+					sock = -1;
+				}
+				/* Purge any remaining messages */
+				k_msgq_purge(&http_data_q);
+				return;
```

Comment:
Give an opportunity to the FSM to be useful. Catch the error and repair it. (ditto for wifi disconnection)